SSA/Projects/CSE/CSE.lean

```diff
@@ -0,0 +1,526 @@
+      | .isFalse neq => .isFalse (fun CONTRA => by rcases CONTRA; . contradiction)
+
+def Expr.decEq (e e' : Expr Op Γ α) : Decidable (e = e') :=
+  match e, e' with
+  | .mk op ty_eq args regArgs, .mk op' ty_eq' args' regArgs' =>
+    if OP : op = op'
+    then
+      match argVector.decEq args (OP ▸ args') with
+      | .isTrue ARGEQ =>
+        match regionVector.decEq _ regArgs (OP ▸ regArgs') with
+        | .isTrue REGEQ =>
+            .isTrue (by subst ty_eq; subst OP; subst ARGEQ; subst REGEQ; rfl; done)
+        | .isFalse neq =>
+          .isFalse (fun CONTRA => by rcases CONTRA; . contradiction)
+      | .isFalse neq =>
+        .isFalse (fun CONTRA => by rcases CONTRA; . contradiction)
+    else .isFalse (fun CONTRA => by rcases CONTRA; . contradiction)
+
+def Com.decEq (c c' : Com Op Γ α) : Decidable (c = c') :=
+  match c, c' with
+  | .ret v, .ret v' =>
+    have VAR_DECEQ : DecidableEq (Ctxt.Var Γ α) := inferInstance
+    match VAR_DECEQ v v' with
+    | .isTrue heq => .isTrue (by simp [heq])
+    | .isFalse f => .isFalse (fun CONTRA => by rcases CONTRA; . contradiction)
+  | .lete (α := α) e body, .lete (α := β) e' body' =>
+    match decEq α β with
+    | .isTrue TY_EQ =>
+      match Expr.decEq e (TY_EQ ▸ e') with
+      | .isFalse neq => .isFalse (fun CONTRA => by rcases CONTRA; . contradiction)
+      | .isTrue EXPR_EQ =>
+        match Com.decEq body (TY_EQ ▸ body') with
+        | .isFalse neq =>  .isFalse (fun CONTRA => by rcases CONTRA; . contradiction)
+        | .isTrue BODY_EQ =>
+          .isTrue (by subst TY_EQ; subst EXPR_EQ; subst BODY_EQ; rfl)
+    | .isFalse f => .isFalse (fun CONTRA => by rcases CONTRA; . contradiction)
+  | .ret .., .lete .. => .isFalse Com.noConfusion
+  | .lete .., .ret .. => .isFalse Com.noConfusion
+end
+termination_by
+Com.decEq c c' =>  sizeOf c
+Expr.decEq e e' => sizeOf e
+regionVector.decEq as bs => sizeOf as
+
+end DecEqCom
+
+namespace CSE
+
+/-- State stored by CSE pass. -/
+structure State (Op : Type) [Goedel Ty] [OpSignature Op Ty] [OpDenote Op Ty] (lets : Lets Op Γstart Γ) where
+  /-- map variable to its canonical value -/
+  var2var : (v : Γ.Var α) → { v' : Γ.Var α // ∀ (V : Γstart.Valuation), (lets.denote V) v = (lets.denote V) v' }
+  /-- map an Expr to its canonical variable -/
+  expr2cache : (α : Ty) → (e : Expr Op Γ α) → Option ({ v : Γ.Var α // ∀ (V : Γstart.Valuation), (lets.denote V) v = e.denote (lets.denote V) })
+
+/-- The empty CSEing state. -/
+def State.empty [Goedel Ty] [OpSignature Op Ty] [OpDenote Op Ty] (lets : Lets Op Γstart Γ) : State Op lets where
+  var2var := fun v => ⟨v, by intros V; rfl⟩
+  expr2cache := fun α e => .none
+
+def State.snocNewExpr2Cache
+ [Goedel Ty] [DecidableEq Ty] [DecidableEq Op] [OpSignature Op Ty] [OpDenote Op Ty]
+ {Γ : Ctxt Ty} {α : Ty}
+ {lets : Lets Op Γstart Γ}
+ (s : State Op lets) (e : Expr Op Γ α) : State Op (Lets.lete lets e) :=
+ {
+  var2var := fun v => by
```

Comment:
I think these are more readable when they are separated (the proofs and the functionality, that is).

Reply:
I am not sure what you mean, it's a dependently typed object, it carries both the functionality and the proof?

Reply:
@bollu sure but dependently typed objects still allow you to use previously-defined things :wink: I mean I don't particularly like that definition of State because it mixes things together. But you could define a version of `snocNewExpr2Cache` that does not build a `State Op _`, but rather builds two functions `var2var :  Γ.Var α →  Γ.Var α ` and `(α : Ty) → (e : Expr Op Γ α) → Option (Γ.Var α)` and then prove that these invariants hold for the two, and then just join everything together in `State.snocNexExpr2Cache`. Does that make sense?